hazelcast/src/main/java/com/hazelcast/sql/impl/AbstractSqlResult.java

```diff
@@ -17,15 +17,51 @@
 package com.hazelcast.sql.impl;
 
 import com.hazelcast.sql.SqlResult;
+import com.hazelcast.sql.SqlRow;
+
+import javax.annotation.Nonnull;
+import java.util.Iterator;
 
 public abstract class AbstractSqlResult implements SqlResult {
 
     public abstract QueryId getQueryId();
 
     public abstract void closeOnError(QueryException exception);
 
+    @Nonnull @Override
+    public abstract ResultIterator<SqlRow> iterator();
+
     @Override
     public void close() {
         closeOnError(QueryException.cancelledByUser());
     }
+
+    public interface ResultIterator<T> extends Iterator<T> {
+
+        /**
+         * A result value from {@link #hasNextImmediately()} meaning that a next
+         * item is available immediately.
+         */
+        int YES = 2;
```

Comment:
Can we convert int constants to enum? This will make methods relying on these constants easier to understand and refactor.
Also, `ResultIterator` could be a top-level class.

Reply:
fixed